lib/geo/wkb.ex

```diff
@@ -22,24 +22,79 @@ defmodule Geo.WKB do
   what the byte order will be.
   """
   @spec encode!(Geo.geometry(), Geo.endian()) :: binary | no_return
-  defdelegate encode!(geom, endian \\ :xdr), to: Encoder
+  def encode!(geom, endian \\ :xdr) do
+    geom |> Encoder.encode!(endian) |> Base.encode16()
+  end
 
   @doc """
   Takes a Geometry and returns a WKB string. The endian decides
   what the byte order will be.
   """
   @spec encode(binary, Geo.endian()) :: {:ok, binary} | {:error, Exception.t()}
-  defdelegate encode(geom, endian \\ :xdr), to: Encoder
+  def encode(geom, endian \\ :xdr) do
+    {:ok, geom |> Encoder.encode!(endian) |> Base.encode16()}
+  rescue
+    exception ->
+      {:error, exception}
+  end
+
+  @doc """
+  Takes a Geometry and returns a WKB string as iodata. The endian decides
+  what the byte order will be.
+  """
+  @spec encode_iodata!(Geo.geometry(), Geo.endian()) :: binary | no_return
+  def encode_iodata!(geom, endian \\ :xdr) do
+    geom |> Encoder.encode!(endian)
+  end
+
+  @doc """
+  Takes a Geometry and returns a WKB string as iodata. The endian decides
+  what the byte order will be.
+  """
+  @spec encode_iodata(binary, Geo.endian()) :: {:ok, binary} | {:error, Exception.t()}
+  def encode_iodata(geom, endian \\ :xdr) do
+    {:ok, geom |> Encoder.encode!(endian)}
+  rescue
+    exception ->
+      {:error, exception}
+  end
+
+  @doc """
+  Takes a WKB as iodata and returns a Geometry.
+  """
+  @spec decode_iodata(iodata()) :: {:ok, Geo.geometry()} | {:error, Exception.t()}
+  def decode_iodata(wkb) do
```

Comment:
I'd avoid decoding iodata (apologies for suggesting it in the first place!) as you'd almost never actually get an iodata to decode. For example, you'd always get a binary from the socket. It'd be complicated to actually decode an iodata so you'd likely first `IO.iodata_to_binary(iodata)` it and decode the binary which is much easier to do, but then we're not _really_ working with iodata.

Ideally this function would have been called simply `decode(wkb)` (of course we can't change it due to backwards compatibility) so my second suggestion is calling it `decode_binary`.

Another suggestion is deprecating `Geo.WKB` module in favour of a new public API:

```
Geo.encode_wkb_to_iodata/2
Geo.decode_wkb/1
Geo.decode_wkb!/1
```

Reply:
Let me try and see if pattern matching will help to where it can simply be `decode(wkb)`

Reply:
interesting!

when it's base16 encoded you're guaranteed it will always start with "00" (`Base.encode16(<<0>>)` or `"01"` so you can totally pattern match on `"00" <> rest` (same as `<<48, 48>>` <> rest`) and it will never overlap with the raw binary which is `<<0>> <> rest` or `<<1>> <> rest`. Clever :D

Reply:
Thanks! And looks like it should work out